sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala

```diff
@@ -114,12 +114,24 @@ class ParquetFilterSuite extends QueryTest with ParquetTest with SharedSQLContex
             new SparkToParquetSchemaConverter(conf).convert(df.schema), pred)
           assert(maybeFilter.isDefined, s"Couldn't generate filter predicate for $pred")
           // Doesn't bother checking type parameters here (e.g. `Eq[Integer]`)
-          maybeFilter.exists(_.getClass === filterClass)
+          assert(flattenPredicateTree(maybeFilter.get).exists(_.getClass === filterClass))
         }
         checker(stripSparkFilter(query), expected)
     }
   }
 
+  private def flattenPredicateTree(filterPredicate: FilterPredicate): Seq[Object] = {
```

Comment:
Hm, but I guess we're a bit unclear if we're going to check nested filters or only top filter.

Hmmm .. I made another take based upon your branch (see https://github.com/nandorKollar/spark/pull/1) by disabling problematic rule(https://github.com/apache/spark/commit/ef77003178eb5cdcb4fe519fc540917656c5d577) explicitly. You could merge that into your branch so that it can be pushed to your PR.

If we find no better way, it should be okay to disable that rule for this test specifically since technically it had to be done together at https://github.com/apache/spark/commit/ef77003178eb5cdcb4fe519fc540917656c5d577.

Reply:
Ok, we can do this too, I like this idea. Actually after flattening the tree, it is easy to check for the entire predicate being pushed to Parquet: the tests should pass all predicates as a flattened tree as expected value in a sequence, and compare this with the flattened actual predicate tree. Most of the cases we are interested in only one predicate, so null filter rule just produces some noise in this case, however in cases like [this](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala#L182) it might provide some more value to check for the entire predicate tree, not just for the presence of 'or'.

Anyway, this requires a lot more change though, I'm fine with your proposed solution!